ImageStreamIO.c

```diff
@@ -1097,28 +1172,70 @@ errno_t ImageStreamIO_createIm_gpu(
         }
         image->md->shared = 0;
         image->md->inode = 0;
-        if (image->md->NBkw > 0)
-        {
-            image->kw = (IMAGE_KEYWORD *)malloc(sizeof(IMAGE_KEYWORD) * image->md->NBkw);
-            if (image->kw == NULL)
-            {
-                printf("Memory allocation error %s %d\n", __FILE__, __LINE__);
-                abort();
-            }
-        }
-        else
-        {
-            image->kw = NULL;
+    }
+
+    ////////////////////////////////////////////////////////////////
+    // Load IMAGE_METADATA struct at address map with all parameters
+    // and calculate sizes and address image-> pointers
+    ////////////////////////////////////////////////////////////////
+    errno_t ierrno = ImageStreamIO_image_sizing_from_scratch(
+                         image, name, naxis, size, datatype
+                         , location, shared, NBsem, NBkw
+                         , imagetype, CBsize, map
+                     );
+    if (ierrno != IMAGESTREAMIO_SUCCESS) {
+        return ierrno;
+    }
+
+    if (shared == 1)
+    {
+        // - Store the inode of the shmim flle into image->md->inode
+        //   - On error, shmim will have been closed; return
+        ierrno = ImageStreamIO_store_image_inode(image);
+        if (ierrno != IMAGESTREAMIO_SUCCESS) {
+            return ierrno;
         }
-        image->md->CBsize = 0;
-        image->md->CBindex = 0;
-        image->md->CBcycle = 0;
+    }
+
+    if(shared != 1)
+    {
+        strncpy(image->name, image->md->name, STRINGMAXLEN_IMAGE_NAME); // local name
+        // Ensure image and image metadata names are null-terminated
+        image->name[STRINGMAXLEN_IMAGE_NAME-1] = '\0';
+    }
+
+    image->md->creatorPID = getpid();
+    image->md->ownerPID = 0; // default value, indicates unset
+
+    image->md->CBsize = 0;
+    image->md->CBindex = 0;
+    image->md->CBcycle = 0;
 
 #ifdef IMAGESTRUCT_WRITEHISTORY
-        image->md->wCBindex = 0;
-        image->md->wCBcycle = 0;
+    image->md->wCBindex = 0;
+    image->md->wCBcycle = 0;
 #endif
+
+
+
+
+
+    if (image->md->NBkw > 0)
+    {
+        image->kw = (IMAGE_KEYWORD *)malloc(sizeof(IMAGE_KEYWORD) * image->md->NBkw);
+        if (image->kw == NULL)
+        {
+            printf("Memory allocation error %s %d\n", __FILE__, __LINE__);
+            abort();
+        }
```

Comment:
For shared, non-local memory, keywords are in the shmim, so this malloc should only be performed when the memory is local, so move this back into the **else** clause of **if (shared == 1) {...} else**.